### PR TITLE
Try with resources

### PR DESCRIPTION
**Pull request by tmy1313**

### CODE REVIEW

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/job/MeaningfulUseUploadJob.java [hunk 1]
```diff
@@ -141,11 +133,16 @@ public void run() {
             LOGGER.error(msg);
             addJobMessage(msg);
             updateStatus(100, JobStatusType.Error);
+        } finally {
             try {
-                parser.close();
+                if (parser != null && !parser.isClosed()) {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/job/MeaningfulUseUploadJob.java]: I thought we were going to use the "try with resources" thing. So lines 55/56 should be put into the `try (BuffererdReader = new...` format

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProductDownloadableResourceCreatorJob.java [hunk 2]
```diff
@@ -63,53 +62,54 @@ public void execute(final JobExecutionContext jobContext) throws JobExecutionExc
...
+                    CertifiedProductCsvPresenter csvPresenter = getCsvPresenter()) {
+
+                initializeWritingToFiles(xmlPresenter, csvPresenter);
+
+                List<CompletableFuture<CertifiedProductSearchDetails>> futures =
+                        new ArrayList<CompletableFuture<CertifiedProductSearchDetails>>();
+    
+                for (CertifiedProductDetailsDTO dto : listings) {
+                    CompletableFuture<CertifiedProductSearchDetails> cpCompletableFuture =
+                        CompletableFuture.supplyAsync(new Supplier<CertifiedProductSearchDetails>() {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProductDownloadableResourceCreatorJob.java]: Can this  new Supplier definition be moved out to be a separate class? I did something like that with some Comparators in the chpl-api project to make them static  at findbug's suggestion; not sure if that's possible here

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProductDownloadableResourceCreatorJob.java [hunk 3]
```diff
@@ -122,23 +122,17 @@ public void accept(final CertifiedProductSearchDetails cp) {
                 edition);
     }
 
-    private void initializeWritingToFiles() throws IOException {
-        xmlPresenter = new CertifiedProductXmlPresenter();
+    private void initializeWritingToFiles(CertifiedProductXmlPresenter xmlPresenter, 
+            CertifiedProductCsvPresenter csvPresenter) throws IOException {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProductDownloadableResourceCreatorJob.java]: Should these be `final`? Checkstyle likely thinks so

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/summarystatistics/StatsCsvFileWriter.java [hunk 4]
```diff
@@ -31,66 +30,44 @@
...
-                statRecord.add(String.valueOf(stat.getTotal2011Listings()));
-                statRecord.add(String.valueOf(stat.getTotalSurveillanceActivities()));
-                statRecord.add(String.valueOf(stat.getTotalOpenSurveillanceActivities()));
-                statRecord.add(String.valueOf(stat.getTotalClosedSurveillanceActivities()));
-                statRecord.add(String.valueOf(stat.getTotalNonConformities()));
-                statRecord.add(String.valueOf(stat.getTotalOpenNonconformities()));
-                statRecord.add(String.valueOf(stat.getTotalClosedNonconformities()));
-                csvFilePrinter.printRecord(statRecord);
+            try (FileWriter fileWriter = new FileWriter(fileName);
+                    CSVPrinter csvFilePrinter = new CSVPrinter(fileWriter, csvFileFormat);) {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/summarystatistics/StatsCsvFileWriter.java]: Not sure if the last `;` is needed in the list. I don't have it in any of mine, and think it looks odd...

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/CertifiedProductCsvPresenter.java [hunk 5]
```diff
@@ -26,7 +26,7 @@
  * @author alarned
  *
  */
-public class CertifiedProductCsvPresenter implements CertifiedProductPresenter {
+public class CertifiedProductCsvPresenter implements CertifiedProductPresenter, AutoCloseable {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/CertifiedProductCsvPresenter.java]: Does AutoCloseable not need anything else added to it?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/Sed2015CsvPresenter.java [hunk 6]
```diff
@@ -31,25 +31,20 @@
...
-
-            csvPrinter.printRecord(generateHeaderValues());
-
-            for (CertifiedProductSearchDetails currListing : cpList) {
-                List<List<String>> rows = generateRows(currListing);
-                if (rows != null) { // can return null to skip a row
-                    for(List<String> row : rows) {
-                        csvPrinter.printRecord(row);
-                        numRows++;
+            try (OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8);
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/Sed2015CsvPresenter.java]: Wait, is this a `try { try () { } }` structure? With nothing in the first try?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductUploadHandlerFactoryImpl.java [hunk 7]
```diff
@@ -83,21 +83,21 @@ public CertifiedProductUploadHandler getHandler(CSVRecord heading, List<CSVRecor
...
-            //would not match any of the options in the db)
-            writer = new CSVPrinter(buf, CSVFormat.EXCEL.withRecordSeparator(""));
-            writer.printRecord(trimmedHeaderVals);
+            try (CSVPrinter writer = new CSVPrinter(buf, CSVFormat.EXCEL.withRecordSeparator(""))) {
+                //adding a blank char as the record separator prevents an \r\n char 
+                //from getting appended to the end of the string (\r\n at the end 
+                //would not match any of the options in the db)
+                writer.printRecord(trimmedHeaderVals);
+            }
         } catch(IOException ex) {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductUploadHandlerFactoryImpl.java]: `catch(` -> `catch (`

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/util/CertificationResultRules.java [hunk 8]
```diff
@@ -63,42 +54,46 @@ public CertificationResultRules() {
...
-                            }else{
-                            	option.setCanHaveOption(canHaveProperty);
-                            }
-                            if (rules.get(certNumber) == null) {
-                                List<CertificationResultOption> options = new ArrayList<CertificationResultOption>();
-                                options.add(option);
-                                rules.put(certNumber, options);
-                            } else {
-                                rules.get(certNumber).add(option);
+            try (InputStream xmlInput = cpr.getInputStream()) {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/util/CertificationResultRules.java]: A double try again?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/util/FileUtils.java [hunk 9]
```diff
@@ -15,17 +15,17 @@
...
+        //read the file into a string
         StringBuffer data = new StringBuffer();
-        BufferedReader reader = null;
         try {
-            reader = new BufferedReader(new InputStreamReader(file.getInputStream()));
-            String line = null;
-            while ((line = reader.readLine()) != null) {
-                if (data.length() > 0) {
-                    data.append(System.getProperty("line.separator"));
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(file.getInputStream()))) {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/util/FileUtils.java]: Double try?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProductDownloadableResourceCreatorJob.java (hunk 3 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProductDownloadableResourceCreatorJob.java]: Change made

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/summarystatistics/StatsCsvFileWriter.java (hunk 4 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/summarystatistics/StatsCsvFileWriter.java]: Fixed

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/CertifiedProductCsvPresenter.java (hunk 5 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/CertifiedProductCsvPresenter.java]: The CertifiedProductPresenter interface already had a close method.  I removed it from that interface, and added the AutoCloseable interface to the class.  Basically, I just moved where the cloes() interface method was defined.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/Sed2015CsvPresenter.java (hunk 6 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/Sed2015CsvPresenter.java]: I had a mental fart...  For some reason, it never dawned on me that that try() could have a catch.  Not sure what I was thinking.  
Fixed.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductUploadHandlerFactoryImpl.java (hunk 7 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductUploadHandlerFactoryImpl.java]: Fixed.

Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductUploadHandlerFactoryImpl.java]: Fixed.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/util/CertificationResultRules.java (hunk 8 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/util/CertificationResultRules.java]: Fixed.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/util/FileUtils.java (hunk 9 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/util/FileUtils.java]: Fixed.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/job/MeaningfulUseUploadJob.java (hunk 1 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/job/MeaningfulUseUploadJob.java]: I accidentally modified this file.  This is something that Katy is working on.  I'm going to revert my changes.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProductDownloadableResourceCreatorJob.java (hunk 2 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/CertifiedProductDownloadableResourceCreatorJob.java]: Change made

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductUploadHandlerFactoryImpl.java (hunk 7 above)
Comment (andlar) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductUploadHandlerFactoryImpl.java]: I don't see the fix...

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/SurveillanceCsvPresenter.java [hunk 10]
```diff
@@ -51,13 +50,9 @@ public SurveillanceCsvPresenter(final Properties props) {
      * @param cpList list of Certified Products
      */
     public void presentAsFile(final File file, final List<CertifiedProductSearchDetails> cpList) {
-        FileWriter writer = null;
-        CSVPrinter csvPrinter = null;
-        try {
-            writer = new FileWriter(file);
-            csvPrinter = new CSVPrinter(writer, CSVFormat.EXCEL);
+        try (FileWriter writer = new FileWriter(file);
+                CSVPrinter csvPrinter = new CSVPrinter(writer, CSVFormat.EXCEL);) {
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/presenter/SurveillanceCsvPresenter.java]: About half of your multi-resource trys have a semi-colon after the last resource (like this one) and half don't. I'm okay with either way, but we need to pick one and stick with it. I'd prefer the version without the trailing semi-colon, and that's the way I did my trys, but could be persuaded to go the other way with a good reason

**Pull request merged**